iothub/device/src/Transport/Amqp/AmqpUnit.cs

```diff
@@ -339,20 +295,109 @@ public async Task<Message> ReceiveMessageAsync(TimeSpan timeout)
             }
             finally
             {
-                if (Logging.IsEnabled)
+                Logging.Exit(this, timeout, $"{nameof(ReceiveMessageAsync)}");
+            }
+        }
+
+        public async Task EnableReceiveMessageAsync(TimeSpan timeout)
+        {
+            if (_closed)
+            {
+                throw new IotHubException("Device is now offline.", false);
+            }
+
+            Logging.Enter(this, timeout, $"{nameof(EnableReceiveMessageAsync)}");
+
+            try
+            {
+                // Wait to grab the semaphore, and then open the telemetry receiving link and set the callback,
+                // and set _isDeviceReceiveMessageCallbackSet  to true.
+                // Once _isDeviceReceiveMessageCallbackSet  is set to true, all received c2d messages will be returned on the callback,
+                // and not via the polling ReceiveAsync() call.
+                bool gain = await _messageReceivingCallbackSemaphore.WaitAsync(timeout).ConfigureAwait(false);
+                if (!gain)
```

Comment:
I don't know what gain means. gainedSemaphore? receivedSemaphore?

Reply:
yeah, gained semaphore. I can rename this.

Reply:
`enteredSemaphore`